### ISSUE TITLE
YetiFrozen

### ISSUE THREAD

**Issue by Rspigler** [JWWeatherman/yeticold#82]
YetiFrozen will be a multi-signature, multi-device, offline, persistent wallet (with one additional online computer for transaction broadcast).

It will be similar to YetiCold, with the added benefit that instead of having all keys generated on one computer (and then deleting that device to get rid of the single point of attack), the keys will be generated separately on each offline computer.  Spending then no longer requires gathering the paper backups and recovering the wallet (on a single device).  The downside is in the added required cost.

BIP 45 support (https://github.com/bitcoin/bips/blob/master/bip-0045.mediawiki):  Both the Master private keys and master public keys are generated independently and not shared with cosigners.  The next index (hardened purpose extended public key) is shared.  Sorting this lexicographically allows all cosigners to independently derive all the same addresses, without any further communication.

YetiFrozen+:  Will include notes/recommendations in the instructions/docs for extra security.  This will include destroying the printer used, running some of the code on OpenPower machines, and using multiple OS's (describing the rational for these decisions should be included in the docs).

Thoughts on OSes:
	Qubes, run on Whonix ws and gw:	Minimization of TCB due to Xen (Type 1 Hypervisor), DNS leaks are impossible (even malware with root privileges can't discover the true IP address (as long as there's no exploit against Xen or Tor)), Best default hardening (based on Kicksecure (Linux Kernel Runtime Guard, all settings recommended by the KSPP, App Armor, Hardened Malloc, etc)).  Some special configurations will be needed to prevent 'Tor over Tor'.
	GNU Guix is an OS that is working on reducing the 'trusting trust' attack by minimizing the size of the binary seed bootstrap, and working towards a universal, full source bootstrap.  There currently is no bootstrappable and reproducible OS, but this will likely be the first OS that will be.
	Debian:	A major linux distribution, that is community driven with free software principles, with lots of eyes on code, that can run on OpenPower ISA 

This should all be worked on after the general code cleanup and improvements that are already being worked on

**Comment by Rspigler** [comment]
> What do you mean by this?
While YetiCold is offline multisignature, it is single device (for private key generation, and transaction signing; there is a second online device for broadcast).  To help get rid of this single point of attack, we instruct users to wipe the device after key generation, and then distribute the paper keys.  This requires users to gather the paper keys and recover the wallet on the single offline device to sign, every time they want to spend.  With YetiFrozen, private key generation is done on each individual offline device, so wiping is not required and the wallet can be persistent.  Signing would then be done by passing PSBTs across the offline computers.

> Destroying the printer
It is true that private keys are not printed, but the descriptor could be saved in memory, which would be a privacy risk as it would allow an attacker to derive all addresses.

> Terminal only option
That's a good point about utxos in terminal.  I am open to other suggestions.  My concern was in regards to the attack surface of running Yeti in a browser compared to something like QT, but that writing this for QT would be a huge cost in terms of development hours (for us), while terminal would help with the attack surface without the development cost.

**Comment by gaetawoo** [comment]
> With YetiFrozen, private key generation is done on each individual offline device, so wiping is not required and the wallet can be persistent. Signing would then be done by passing PSBTs across the offline computers.

Then you haven't really benefited from the multisignature nature... because you have all your systems 1. live/accessible. 2. with you in your house. 3. this is basically a more inconvenient version of yeticold without any of the benefits.... unless these devices are stores in various locations off-site which is yeticold but more impractical. I'm not seeing the benefit. I thought Frozen was supposed to be yeticold but each of the 7 keys is made on different machines. The thing is... you'll have to pass around all their xpubs in order to make the multisig descriptor for the multisig wallet. But then after that, it becomes yeticold... the frozen is just different machines for key generation.

> Browser

We could download Chromium for a least trustable, most verifyable browser. Vanilla chromium I cannot imagine would be an attack vector for public keys.

**Comment by Rspigler** [comment]
I definitely disagree.  Multisignature is /supposed/ to be multi-device.  That is industry standard.  Generating keys and signing transactions on a single device (what we do currently) is being a bit generous with the definition of multisignature (IMO).

> 1. live/accessible

The devices with private keys are all offline, shut down when not in use, and distributed geographically (also an industry standard advice).  Obviously, no computer is secure once an adversary obtains access (3rd Immutable Law of Security)- but it will take more time, effort, and money than a piece of paper.

> 2. with you in your house

The computers are to be distributed just like we instruct users to do currently with the paper keys

> this is basically a more inconvenient version of yeticold without any of the benefits..

We've discussed the many benefits? I also believe it is much more usable.  Currently, our users have to (1) redownload Bitcoin Core, (2) recover a quorum of their paper keys from geographically distributed places, (3) recover their wallets every time they want to spend Bitcoin, and (4) then wipe again.  With YetiFrozen, 1,3 and 4 is no longer necessary.  2 is still necessary, but instead of collecting paper keys, users will be turning on laptops.

**Comment by gaetawoo** [comment]
I get the multi-device thing. Understood. But that means, as you described, that you have to have all those devices safely stored in different locations... now it's not just a small little envelope but a freaking computer. Which... maybe that isn't a big deal if you are protecting equivalents of millions, the added cost/trouble/impracticality of storing 7 computers in various geographic locations could be achieved. 

As far as more usable, i mostly disagree. "redownloading Bitcoin Core" is no longer going to be necessary for yeticold or frozen because of the PSBT function. So that's equal. Recovering a quorum is also the same as you said. (3) will still have the issue of not having an up to date blockchain in either case, so the same solution would apply to both. Sure, the wiping again would be a difference. 

Yea I think my main point is not that it shouldn't be an option, but that it's not a more convenient solution than Cold. There are ways it's better (keys stay separate always) and worse (device cost, storage of entire device and accessories needed to run it (power cable etc)).

**Comment by Rspigler** [comment]
> added cost/trouble/impracticality of storing 7 computers in various geographic locations...

Yes, I think it definitely is for more money than amounts designed for Yeticold/warm/hot.  Personally, I would store less than millions of dollars in a YetiFrozen setup, and I don't know if it should be up to us to designate explicit monetary cutoffs between the levels besides describing the pros and cons like outlined to do in #81 

How I like to visualize this, is that it is the same as standards recommend today, but without the security theatre and negative consequences of using HWWs - replacing them with laptops instead.  Obviously, laptops are larger and more expensive than HWWs.

> redownloading Bitcoin Core is no longer going to be necessary for yeticold or frozen because of the PSBT function

Well, for YetiFrozen yes, redownloading will not be necessary - that was my argument.  Even with PSBTs, users will still have to redownload Bitcoin Core with YetiCold, just like they have to reinstall Ubuntu (since they are wiping the computer after each use).  The point is, the wallet is not persistent, and the users are having to recover the wallet each time they want to spend any bitcoin.  This is a big hassle, and unique only to Yeti, as far as I know.  What PSBTs help with, is that they don't have to wait for the blockchain to sync.  Yetifrozen helps with all of this.

> (3) will still have the issue of not having an up to date blockchain in either case

With PSBTs, the offline wallets don't need a blockchain to sign transactions.  If the online computer, which is only running Bitcoin/Yeti, is stored in a secure location (home with no roommates, etc), I don't see why it can't be left running in order to keep up to date.  YetiFrozen users shouldn't have any issue with this. (Open to more debate/analysis here).

> I think my main point is not that it shouldn't be an option, but that it's not a more convenient solution than Cold
(Thumbs up)!
